controllers/hostdefinition_controller.go

```diff
@@ -0,0 +1,456 @@
+
+	for _, cr := range clusterRoles {
+		found := &rbacv1.ClusterRole{}
+		err := r.Get(context.TODO(), types.NamespacedName{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		}, found)
+		if err != nil && errors.IsNotFound(err) {
+			logger.Info("Creating a new ClusterRole", "Name", cr.GetName())
+			err = r.Create(context.TODO(), cr)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			logger.Error(err, "Failed to get ClusterRole", "Name", cr.GetName())
+			return err
+		} else {
+			err = r.Update(context.TODO(), cr)
+			if err != nil {
+				logger.Error(err, "Failed to update ClusterRole", "Name", cr.GetName())
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *HostDefinitionReconciler) deleteClusterRoles(instance *hostdefinition.HostDefinition) error {
+	logger := hostDefinitionLog.WithName("deleteClusterRoles")
+
+	clusterRoles := r.getClusterRoles(instance)
+
+	for _, cr := range clusterRoles {
+		found := &rbacv1.ClusterRole{}
+		err := r.Get(context.TODO(), types.NamespacedName{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		}, found)
+		if err != nil && errors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logger.Error(err, "failed to get ClusterRole", "Name", cr.GetName())
+			return err
+		} else {
+			logger.Info("deleting ClusterRole", "Name", cr.GetName())
+			if err := r.Delete(context.TODO(), found); err != nil {
+				logger.Error(err, "failed to delete ClusterRole", "Name", cr.GetName())
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *HostDefinitionReconciler) getClusterRoles(instance *hostdefinition.HostDefinition) []*rbacv1.ClusterRole {
+	hostdefinition := instance.GenerateHostDefinitionClusterRole()
+
+	return []*rbacv1.ClusterRole{
+		hostdefinition,
+	}
+}
+
+func (r *HostDefinitionReconciler) reconcileServiceAccount(instance *hostdefinition.HostDefinition) error {
+	logger := hostDefinitionLog.WithValues("Resource Type", "ServiceAccount")
+
+	hostDefinition := instance.GenerateServiceAccount()
+
+	for _, sa := range []*corev1.ServiceAccount{
+		hostDefinition,
+	} {
+		if err := controllerutil.SetControllerReference(instance.Unwrap(), sa, r.Scheme); err != nil {
+			return err
+		}
+		found := &corev1.ServiceAccount{}
+		err := r.Get(context.TODO(), types.NamespacedName{
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		}, found)
+		if err != nil && errors.IsNotFound(err) {
+			logger.Info("Creating a new ServiceAccount", "Namespace", sa.GetNamespace(), "Name", sa.GetName())
+			err = r.Create(context.TODO(), sa)
+			if err != nil {
+				return err
+			}
+
+			rErr := r.restartDeployment(logger, instance)
+			if rErr != nil {
+				return rErr
+			}
+
+		} else if err != nil {
+			logger.Error(err, "Failed to get ServiceAccount", "Name", sa.GetName())
+			return err
+		} else {
+			// Resource already exists - don't requeue
+			//logger.Info("Skip reconcile: ServiceAccount already exists", "Namespace", sa.GetNamespace(), "Name", sa.GetName())
```

Comment:
Do we need that comments?

Reply:
I don't think so, but it wasn't a new code, do you want me to handle it?

Reply:
I copied it from [here](https://github.com/IBM/ibm-block-csi-operator/blob/060bde032636e530a43764b00802cb03f5ec0870/controllers/ibmblockcsi_controller.go#L519)

Reply:
yes, please remove

Reply:
done